### PR TITLE
[Proposal] Implement the typed-function references proposal.

### PR DESCRIPTION
**Pull request by q82419**

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review bot](https://github.com/flows-network/github-pr-summary/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

**Comment by q82419** [comment]
@apepkuss  You can refer to the API header for the changes. I'll notice you once the docs ready. Thanks.

**Comment by apepkuss** [comment]
@q82419 Thanks for the reminder. I'll check the changes.

File: lib/executor/helper.cpp [hunk 1]
```diff
@@ -252,5 +252,63 @@ Executor::getDataInstByIdx(Runtime::StackManager &StackMgr,
...
+                         const ValType &Exp,
+                         const Runtime::Instance::ModuleInstance &ModGot,
+                         const ValType &Got) const noexcept {
+  if (!Exp.isRefType() && !Got.isRefType() && Exp.getCode() == Got.getCode()) {
+    // Match for the non-reference type case.
+    return true;
+  }
+  if (Exp.isRefType() && Got.isRefType()) {
+    // Nullable matching.
+    bool NullableMatch = Exp.isNullableRefType() || !Got.isNullableRefType();
```
Comment (ibmibmibm) [inline lib/executor/helper.cpp:265]:
```c++
if (!Exp.isNullableRefType() && Got.isNullableRefType()) {
 return false;
}
```

File: lib/loader/serialize/serial_type.cpp [hunk 2]
```diff
@@ -6,6 +6,91 @@
...
+    if (unlikely(!Conf.hasProposal(Proposal::ReferenceTypes))) {
+      return logNeedProposal(ErrCode::Value::MalformedElemType,
+                             Proposal::ReferenceTypes, From);
+    }
+    [[fallthrough]];
+  case TypeCode::FuncRef:
+    OutVec.push_back(static_cast<uint8_t>(Code));
+    return {};
+  case TypeCode::TypeIndex:
+    if (!Conf.hasProposal(Proposal::FunctionReferences)) {
```
Comment (ibmibmibm) [inline lib/loader/serialize/serial_type.cpp:25]: unlikely

File: lib/validator/formchecker.cpp [hunk 3]
```diff
@@ -108,66 +114,71 @@ void FormChecker::addElem(const AST::ElementSegment &Elem) {
...
-    assumingUnreachable();
+bool FormChecker::matchType(const ValType &Exp,
+                            const ValType &Got) const noexcept {
+  if (!Exp.isRefType() && !Got.isRefType() && Exp.getCode() == Got.getCode()) {
+    // Match for the non-reference type case.
+    return true;
   }
+  if (Exp.isRefType() && Got.isRefType()) {
+    // Nullable matching.
+    bool NullableMatch = Exp.isNullableRefType() || !Got.isNullableRefType();
```
Comment (ibmibmibm) [inline lib/validator/formchecker.cpp:140]:
```c++
if (!Exp.isNullableRefType() && Got.isNullableRefType()) {
  return false;
}
```

File: lib/validator/formchecker.cpp [hunk 4]
```diff
@@ -545,49 +630,108 @@ Expect<void> FormChecker::checkInstr(const AST::Instruction &Instr) {
...
+    assuming(TypeIdx < Types.size());
+    return StackTrans({}, {ValType(TypeCode::Ref, TypeIdx)});
+  }
+  case OpCode::Ref__as_non_null: {
+    if (auto Res = popType()) {
+      if (*Res == unreachableVType()) {
+        pushType(unreachableVType());
+        return {};
+      }
+      if (!(**Res).isRefType()) {
```
Comment (ibmibmibm) [inline lib/validator/formchecker.cpp:717]: `!(*Res)->isRefType()`

**Review by ibmibmibm — CHANGES_REQUESTED** [review]

**Review by ibmibmibm — APPROVED** [review]

**Pull request merged**